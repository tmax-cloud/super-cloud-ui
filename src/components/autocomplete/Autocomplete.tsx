import * as React from 'react';
import { Autocomplete as MuiAutocomplete, Theme, AutocompleteRenderInputParams, SxProps, AutocompleteInputChangeReason, AutocompleteCloseReason, AutocompleteChangeReason, AutocompleteChangeDetails } from '@mui/material';
import { ListContainer } from './ListContainer';
import { RoundButtonItem } from './RoundButtonItem';

const renderOptionByType = (optionType?: optionType) => {
  const renderLabelOption = (props: React.HTMLAttributes<HTMLLIElement>, option: string) => <RoundButtonItem key={props.id} props={props} item={option} />;
  return optionType === 'label' ? renderLabelOption : undefined;
};

const Autocomplete = (props: AutocompleteProps) => {
  const { optionType, ...rest } = props;
  return <MuiAutocomplete disableClearable freeSolo PaperComponent={ListContainer} renderOption={renderOptionByType(optionType)} {...rest} />;
};

export default Autocomplete;

export type optionType = 'label' | undefined;

export interface AutocompleteProps {
  /**
   * Override the default text for the *clear* icon button.
   * @default 'Clear'
   */
  clearText?: string;
  /**
   * Override the default text for the *close popup* icon button.
   * @default 'Close'
   */
  closeText?: string;
  /**
   * The default value.
   * @default 'null'
   */
  defaultValue?: any;
  /**
   * 	This prop is used to help implement the accessibility logic. If you don't provide an id it will fall back to a randomly generated one.
   */
  id?: string;
  /**
   * The input value.
   */
  inputValue?: string;
  /**
   * Callback fired when the value changes.
   *
   * @param {React.SyntheticEvent} event The event source of the callback.
   * @param {string|string[]} value The new value of the component.
   * @param {string} reason One of "createOption", "selectOption", "removeOption", "blur" or "clear".
   * @param {string} [details]
   */
  onChange?: (event: React.SyntheticEvent, value: any, reason: AutocompleteChangeReason, details?: AutocompleteChangeDetails<string>) => void;
  /**
   * Callback fired when the popup requests to be closed.
   *
   * @param {React.SyntheticEvent} event The event source of the callback.
   * @param {string} reason Can be: `"toggleInput"`, `"escape"`, `"selectOption"`, `"removeOption"`, `"blur"`.
   */
  onClose?: (event: React.SyntheticEvent, reason: AutocompleteCloseReason) => void;
  /**
   * Callback fired when the input value changes.
   *
   * @param {React.SyntheticEvent} event The event source of the callback.
   * @param {string} value The new value of the text input.
   * @param {string} reason Can be: `"input"` (user input), `"reset"` (programmatic change), `"clear"`.
   */
  onInputChange?: (event: React.SyntheticEvent, value: string, reason: AutocompleteInputChangeReason) => void;
  /**
   * Callback fired when the popup requests to be opened.
   * Use in controlled mode (see open).
   *
   * @param {React.SyntheticEvent} event The event source of the callback.
   */
  onOpen?: (event: React.SyntheticEvent) => void;
  /**
   * If `true`, the component is shown.
   */
  open?: boolean;
  /**
   * Array of options.
   */
  options: ReadonlyArray<string>;
  /**
   * Option type. Can be: `"name"`, `"label"`.
   */
  optionType?: optionType;
  /**
   * Render the input.
   *
   * @param {object} params
   * @returns {ReactNode}
   */
  renderInput: (params: AutocompleteRenderInputParams) => React.ReactNode;
  /**
   * The system prop that allows defining system overrides as well as additional CSS styles.
   */
  sx?: SxProps<Theme>;
  /**
   * The value of the autocomplete.
   */
  value?: any;
}
