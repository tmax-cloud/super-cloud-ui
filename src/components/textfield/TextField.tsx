import * as React from 'react';
import { styled, TextField as MuiTextField, Theme } from '@mui/material';
import ThemeWrapper from '../../themes/ThemeWrapper';
import Error from '../../assets/images/error.svg';

const getHoverBorderBottomColor = (theme: Theme, ownerState: TextFieldProps) => {
  if (ownerState.readOnly || ownerState.disabled) {
    return theme.palette.textfield.borderBottom;
  }
  if (ownerState.error) {
    return theme.palette.textfield.errorBorderBottom;
  }
  return theme.palette.textfield.hoverBorderBottom;
};

const StyledTextField = styled(MuiTextField, {
  shouldForwardProp: (prop) => prop !== 'ownerState',
})<{ ownerState: TextFieldProps }>(({ theme, ownerState }) => ({
  '&.MuiTextField-root': {
    width: '100%',
    height: 'calc(1rem * 1.5 + 1px * 2 + calc(0.375rem - 1px) + calc(0.375rem - 1px))',
  },
  '& .MuiOutlinedInput-root': {
    backgroundColor: ownerState.readOnly ? theme.palette.textfield.readonlyBackground : theme.palette.textfield.background,
    border: `1px solid ${theme.palette.textfield.border}`,
    borderBottomColor: ownerState.error ? theme.palette.textfield.errorBorderBottom : theme.palette.textfield.borderBottom,
    borderBottomWidth: ownerState.error ? '2px' : '1px',
    borderRadius: 0,
    backgroundImage: ownerState.error && `url(${Error})`,
    backgroundPosition: ownerState.error && 'calc(100% - 0.5rem) center',
    backgroundSize: ownerState.error && '1rem 1rem',
    backgroundRepeat: ownerState.error && 'no-repeat',
    '& fieldset': {
      border: 'none',
      padding: 0,
    },
    '&:hover': {
      borderBottomColor: getHoverBorderBottomColor(theme, ownerState),
    },
    '&:hover fieldset': {
      border: 'none',
    },
    '& .Mui-disabled': {
      color: theme.palette.textfield.disabled,
      backgroundColor: theme.palette.textfield.disabledBackground,
      cursor: 'not-allowed',
      WebkitTextFillColor: 'initial',
    },
  },
  '& .MuiInputBase-input': {
    padding: `${theme.spaces.textfield.paddingY} ${theme.spaces.textfield.paddingX}`,
    paddingBottom: ownerState.error ? theme.spaces.textfield.errorPaddingBottom : theme.spaces.textfield.paddingY,
    paddingRight: ownerState.error ? theme.spaces.textfield.errorPaddingRight : theme.spaces.textfield.paddingX,
    color: theme.palette.global.color100,
    fontSize: theme.typography.textfield.fontSize,
    lineHeight: 1.5,
    textOverflow: 'ellipsis',
    '&:focus': {
      outline: 'auto',
    },
    '&::placeholder': {
      color: theme.palette.textfield.placeholder,
    },
  },
  '& .MuiFormHelperText-root': {
    marginLeft: 0,
    fontSize: theme.typography.textfield.fontSize,
  },
}));

const TextField = (props: TextFieldProps) => {
  const { readOnly, ...rest } = props;
  return (
    <ThemeWrapper>
      <StyledTextField variant="outlined" ownerState={props} InputProps={{ readOnly }} {...rest} />
    </ThemeWrapper>
  );
};

export default TextField;

export interface TextFieldProps {
  /**
   * Class name applied to the root element.
   */
  className?: string;
  /**
   * The default value. Use when the component is not controlled.
   */
  defaultValue?: unknown;
  /**
   * If `true`, the component is disabled.
   */
  disabled?: boolean;
  /**
   * If `true`, the `input` will indicate an error.
   */
  error?: boolean;
  /**
   * The helper text content.
   */
  helperText?: React.ReactNode;
  /**
   * The id of the `input` element.
   */
  id?: string;
  /**
   * Name attribute of the `input` element.
   */
  name?: string;
  onBlur?: React.FocusEventHandler;
  onClick?: React.MouseEventHandler;
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  onFocus?: React.FocusEventHandler;
  onKeyDown?: React.KeyboardEventHandler<HTMLInputElement>;
  onKeyUp?: React.KeyboardEventHandler<HTMLInputElement>;
  /**
   * The short hint displayed in the `input` before the user enters a value.
   */
  placeholder?: string;
  /**
   * It prevents the user from changing the value of the field
   */
  readOnly?: boolean;
  /**
   * If `true`, the `input` element is required.
   */
  required?: boolean;
  /**
   * Type of the `input` element. It should be [a valid HTML5 input type](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#Form_%3Cinput%3E_types).
   */
  type?: string;
  /**
   * The value of the `input` element, required for a controlled component.
   */
  value?: unknown;
}

TextField.defaultProps = {
  disabled: false,
  error: false,
  readOnly: false,
  required: false,
  type: 'text',
};
