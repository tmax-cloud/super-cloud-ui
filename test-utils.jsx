// import React from 'react';
// import { render } from '@testing-library/react';
// 추후에 redux 추가하게 되면 redux Provider로 감싸주는 custom render 메서드 추가 해야됨.

export * from '@testing-library/react';
