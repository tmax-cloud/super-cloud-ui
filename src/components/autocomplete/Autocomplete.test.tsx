import * as React from 'react';
import { render, screen } from '../../../test-utils';
import TextField from '../textfield/TextField';
import Autocomplete from './Autocomplete';

describe('Autocomplete', () => {
  test('renders correctly', () => {
    const { container } = render(<Autocomplete open options={['one', 'two']} renderInput={(params) => <TextField {...params} />} />);
    expect(container).toMatchSnapshot();
  });
  test('renders correctly if option type is label', () => {
    render(<Autocomplete open options={['one', 'two']} optionType="label" renderInput={(params) => <TextField {...params} />} />);
    const listbox = screen.getByRole('listbox');
    const options = screen.getAllByRole('option');
    // check option size
    expect(options).toHaveLength(7);
    // check listbox contains options
    options.forEach((option) => {
      expect(listbox).toContainElement(option);
    });
    // check button style
    const button = screen.getAllByRole('button')[0];
    expect(button).toHaveStyle('color: #8476d1');
  });
});
