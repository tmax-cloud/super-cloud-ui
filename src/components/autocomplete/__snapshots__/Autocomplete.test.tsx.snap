// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete renders correctly 1`] = `
<div>
  <div
    aria-expanded="true"
    aria-owns="mui-2-listbox"
    class="MuiAutocomplete-root css-16awh2u-MuiAutocomplete-root"
    role="combobox"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1mfi12u-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-controls="mui-2-listbox"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          id="mui-2"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-nnbavb"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
